newlib/newlib-4.2.0.20211231.patch

```diff
@@ -0,0 +1,22 @@
+--- a/libtock-newlib-4.2.0.20211231/headers/ssp/string.h
++++ b/libtock-newlib-4.2.0.20211231/headers/ssp/string.h
+@@ -36,7 +36,7 @@
+
+ __BEGIN_DECLS
+ void *__memcpy_chk(void *, const void *, size_t, size_t);
+-void *__memmove_chk(void *, void *, size_t, size_t);
++void *__memmove_chk(void *, const void *, size_t, size_t);
+ void *__mempcpy_chk(void *, const void *, size_t, size_t);
+ void *__memset_chk(void *, int, size_t, size_t);
+ char *__stpcpy_chk(char *, const char *, size_t);
+
```

Comment:
Why do we need to patch newlib? What's the status of this, have you submitted it upstream?

Reply:
We get warnings without those changes.

Would newlib really not notice those warnings after multiple years? I assumed it's something we're doing differently or something they don't want to fix.

Reply:
That seems like an issue on our side then.

I suspect they don't see the warnings. If we want to keep this we should send a fix upstream. Otherwise this is how nothing ever gets fixed

Reply:
Not wanting to learn a new tool doesn't seem like a good excuse here.

Newlib is a maintained project so we should be sending all fixes and patches upstream. Otherwise we are stuck patching newlib for the foreseeable future. We also don't even know if this is the correct fix, by sending the change upstream we can get feedback from the maintainers about the fix. We shouldn't just punt on the right way to do things because it might be a little harder

Reply:
@bradjc FWIW I can volunteer to relay patches upstream, if you don't want to use mailing lists. Would need to dig into this issue first, though.

Reply:
Please do!

Warning we get:

```
CC        ../../libtock/udp.c
In file included from ../../lib/libtock-newlib-4.3.0.20230120/arm/arm-none-eabi/include/string.h:180,
                 from ../../libtock/udp.c:1:
../../lib/libtock-newlib-4.3.0.20230120/arm/arm-none-eabi/include/ssp/string.h:39:7: warning: mismatch in argument 2 type of built-in function '__memmove_chk'; expected 'const void *' [-Wbuiltin-declaration-mismatch]
   39 | void *__memmove_chk(void *, void *, size_t, size_t);
      |       ^~~~~~~~~~~~~
```

Based on `-Wbuiltin-declaration-mismatch`, the glibc definition:

https://github.com/bminor/glibc/blob/2e0c0ff95ca0e3122eb5b906ee26a31f284ce5ab/include/string.h#L189

```c
extern void *__memmove_chk (void *__dest, const void *__src, size_t __len,
			    size_t __destlen) __THROW;
```

versus newlib:

https://github.com/bminor/newlib/blob/d5dcb484c705a214b30826c82b9fd8bf83772093/newlib/libc/include/ssp/string.h#L39

```c
void *__memmove_chk(void *, void *, size_t, size_t);
```

And I suppose we haven't seen this before because we've been using the headers provided by the user's compiler.